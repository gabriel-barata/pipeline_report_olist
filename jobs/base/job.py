import os
from abc import abstractmethod

from pyspark.sql import DataFrame, SparkSession
from pyspark.sql.functions import col, lit


class Job:
    def __init__(
        self,
        spark: SparkSession = None,
        config: dict[str, str] = None,
        is_test: bool = False,
    ):
        self.config = config or {}
        self._spark = spark
        self._is_test = is_test

    @property
    def ref_date(self) -> str:
        return self.config.get('ref_date')

    @property
    def is_test(self) -> bool:
        return self._is_test

    @property
    def aditional_packages(self) -> str:
        return ','.join(
            [
                'org.apache.hadoop:hadoop-aws:3.3.2',
                'com.amazonaws:aws-java-sdk-bundle:1.12.262',
                'io.delta:delta-spark_2.12:3.1.0',
            ]
        )

    @property
    def spark(self) -> SparkSession:
        if self._spark is not None:
            return self._spark

        builder = SparkSession.builder.appName(self.__class__.__name__)

        if os.environ.get('ENV', '').lower() != 'dev':
            return builder.getOrCreate()

        return (
            builder.master('local[*]')
            .config('spark.jars.packages', self.aditional_packages)
            .config('spark.hadoop.fs.s3a.endpoint', 's3.amazonaws.com')
            .config(
                'spark.hadoop.fs.s3a.access.key',
                os.getenv('AWS_ACCESS_KEY_ID'),
            )
            .config(
                'spark.hadoop.fs.s3a.secret.key',
                os.getenv('AWS_SECRET_ACCESS_KEY'),
            )
            .config(
                'spark.hadoop.fs.s3a.impl',
                'org.apache.hadoop.fs.s3a.S3AFileSystem',
            )
            .config('spark.hadoop.fs.s3a.endpoint', 's3.amazonaws.com')
            .config(
                'spark.sql.extensions',
                'io.delta.sql.DeltaSparkSessionExtension',
            )
            .config(
                'spark.sql.catalog.spark_catalog',
                'org.apache.spark.sql.delta.catalog.DeltaCatalog',
            )
            .config('spark.driver.memory', '6g')
            .getOrCreate()
        )

    def generate_partition_columns(self, df: DataFrame) -> DataFrame:
        return (
            df.withColumn('YEAR', lit(self.ref_date[:4]))
            .withColumn('MONTH', lit(self.ref_date[5:7]))
            .withColumn('DAY', lit(self.ref_date[-2:]))
        )

    def filter_by_ref_date(self, df: DataFrame) -> DataFrame:
        year, month, day = (
            self.ref_date[:4],
            self.ref_date[5:7],
            self.ref_date[-2:],
        )

        return df.filter(
            (col('YEAR') == year)
            & (col('MONTH') == month)
            & (col('DAY') == day)
        )

    @abstractmethod
    def run(self) -> None:
        pass

    def save(
        self,
        df: DataFrame,
        target: str,
        mode: str = 'overwrite',
        file_format: str = 'delta',
    ) -> DataFrame | None:
        if self.is_test:
            return df
        df.write.mode(mode).format(file_format).save(target)
