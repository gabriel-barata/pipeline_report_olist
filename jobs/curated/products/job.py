from pyspark.sql import DataFrame
from pyspark.sql.functions import avg, col, when

from jobs.base.job import Job
from jobs.base.variables import CURATED_BUCKET, RAW_BUCKET


class ProductsJob(Job):
    sources = {
        'products': RAW_BUCKET + 'kaggle/products',
        'order_items': RAW_BUCKET + 'kaggle/order_items',
    }
    target = CURATED_BUCKET + 'products/products'

    def run(self) -> None:
        products = self.spark.read.parquet(self.sources['products'])
        products = self.filter_by_ref_date(products)
        order_items = self.source_order_items

        mean_product_value = order_items.groupBy(col('product_id')).agg(
            avg('price').alias('avg_price')
        )

        q1, q2 = mean_product_value.approxQuantile('avg_price', [0.33, 0.66], 0.01)

        mean_product_value = mean_product_value.withColumn(
            'price_category',
            when(col('avg_price') <= q1, 'baixo')
            .when((col('avg_price') > q1) & (col('avg_price') <= q2), 'medio')
            .otherwise('alto'),
        )

        products = products.join(mean_product_value, on='product_id', how='left')

        products = products.select(
            col('product_id'),
            col('product_category_name'),
            col('product_name_lenght'),
            col('product_description_lenght'),
            col('product_photos_qty'),
            col('product_weight_g'),
            col('product_length_cm'),
            col('product_height_cm'),
            col('price_category'),
            col('YEAR'),
            col('MONTH'),
            col('DAY'),
        )

        (products.write.mode('overwrite').format('delta').save(self.target))

    @property
    def source_order_items(self) -> DataFrame:
        return self.spark.read.parquet(self.sources['order_items']).select(
            col('product_id'),
            col('price'),
        )
