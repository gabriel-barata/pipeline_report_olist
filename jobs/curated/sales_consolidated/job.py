from pyspark.sql import DataFrame
from pyspark.sql.functions import col, count, sum

from jobs.base.job import Job
from jobs.base.variables import CURATED_BUCKET, RAW_BUCKET

SRC_BASE_PATH = RAW_BUCKET + 'kaggle/'


class SalesConsolidatedJob(Job):
    sources = {
        'orders': SRC_BASE_PATH + 'orders/',
        'order_items': SRC_BASE_PATH + 'order_items/',
        'products': SRC_BASE_PATH + 'products/',
        'customers': SRC_BASE_PATH + 'customers/',
    }
    target = CURATED_BUCKET + 'sales/sales_consolidated'

    def run(self) -> None | DataFrame:
        orders = self.filter_by_ref_date(self.source_orders)
        order_items = self.filter_by_ref_date(self.source_order_items)
        products = self.filter_by_ref_date(self.source_products)
        customers = self.filter_by_ref_date(self.source_customers)

        data = (
            orders.join(order_items, on='order_id', how='left')
            .join(products, on='product_id', how='left')
            .join(customers, on='customer_id', how='left')
        )

        data = data.groupBy(
            col('customer_state'), col('product_category_name')
        ).agg(
            sum('price').alias('sales_total'),
            count('price').alias('sales_qty_items_sold'),
        )

        data = data.select(
            col('customer_state').alias('sales_state'),
            col('product_category_name').alias('sales_product_category'),
            col('sales_total').cast('decimal(18,2)'),
            col('sales_qty_items_sold'),
        )

        return self.save(data, self.target)

    @property
    def source_orders(self) -> DataFrame:
        return (
            self.spark.read.parquet(self.sources['orders'])
            .filter(col('order_status') == 'delivered')
            .select(
                col('order_id'),
                col('customer_id'),
                col('order_status'),
            )
        )

    @property
    def source_order_items(self) -> DataFrame:
        return self.spark.read.parquet(self.sources['order_items']).select(
            col('order_id'),
            col('product_id'),
            col('price'),
        )

    @property
    def source_products(self) -> DataFrame:
        return self.spark.read.parquet(self.sources['products']).select(
            col('product_id'),
            col('product_category_name'),
        )

    @property
    def source_customers(self) -> DataFrame:
        return self.spark.read.parquet(self.sources['customers']).select(
            col('customer_id'),
            col('customer_state'),
        )
