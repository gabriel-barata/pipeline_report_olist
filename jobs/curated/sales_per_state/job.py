from pyspark.sql import DataFrame
from pyspark.sql.functions import col, sum

from jobs.base.job import Job
from jobs.base.variables import CURATED_BUCKET, RAW_BUCKET

SRC_BASE_PATH = RAW_BUCKET + 'kaggle/'


class SalesPerStateJob(Job):
    sources = {
        'orders': SRC_BASE_PATH + 'orders/',
        'order_items': SRC_BASE_PATH + 'order_items/',
        'customers': SRC_BASE_PATH + 'customers/',
    }
    target = CURATED_BUCKET + '/sales/sales_per_state'

    def run(self) -> None | DataFrame:
        orders = self.filter_by_ref_date(self.source_orders)
        order_items = self.filter_by_ref_date(self.source_order_items)
        customers = self.filter_by_ref_date(self.source_customers)

        data = orders.join(order_items, on='order_id', how='left').join(
            customers, on='customer_id', how='left'
        )

        data = data.groupBy('customer_state').agg(
            sum('price').alias('sales_total')
        )

        data = data.select(
            col('customer_state').alias('sales_customer_state'),
            col('sales_total').cast('decimal(18,2)'),
        )

        return self.save(data, self.target)

    @property
    def source_orders(self) -> DataFrame:
        return (
            self.spark.read.parquet(self.sources['orders'])
            .filter(col('order_status') == 'delivered')
            .select(
                col('order_id'),
                col('customer_id'),
                col('order_status'),
            )
        )

    @property
    def source_order_items(self) -> DataFrame:
        return self.spark.read.parquet(self.sources['order_items']).select(
            col('order_id'),
            col('product_id'),
            col('price'),
        )

    @property
    def source_customers(self) -> DataFrame:
        return self.spark.read.parquet(self.sources['customers']).select(
            col('customer_id'),
            col('customer_state'),
        )
