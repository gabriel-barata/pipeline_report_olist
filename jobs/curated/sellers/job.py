from jobs.base.job import Job
from jobs.base.variables import CURATED_BUCKET, RAW_BUCKET


class SellersJob(Job):
    sources = {'': ...}
    target = ''

    def run(self) -> None:
        ...
