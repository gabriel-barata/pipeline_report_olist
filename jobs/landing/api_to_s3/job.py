import requests
from pyspark.sql.types import StringType, StructField, StructType

from jobs.base.job import Job
from jobs.base.variables import LANDING_BUCKET


class ApiToS3Job(Job):
    target = LANDING_BUCKET + 'api_CNAE/'

    def run(self) -> None:
        data = self.spark.createDataFrame(
            self.fetch_api_data(), schema=self.schema
        )

        data = self.generate_partition_columns(data)

        (
            data.write.mode('overwrite')
            .partitionBy('YEAR', 'MONTH', 'DAY')
            .json(self.target)
        )

    def fetch_api_data(self) -> dict:
        response = requests.get(self.url)
        response.raise_for_status()

        return response.json()

    @property
    def url(self) -> str:
        return 'https://servicodados.ibge.gov.br/api/v2/cnae/classes'

    @property
    def schema(self) -> StructType:
        return StructType(
            [
                StructField('id', StringType(), True),
                StructField('descricao', StringType(), True),
                StructField('observacoes', StringType(), True),
                StructField(
                    'grupo',
                    StructType(
                        [
                            StructField('id', StringType(), True),
                            StructField('descricao', StringType(), True),
                            StructField('observacoes', StringType(), True),
                            StructField(
                                'divisao',
                                StructType(
                                    [
                                        StructField('id', StringType(), True),
                                        StructField(
                                            'descricao', StringType(), True
                                        ),
                                        StructField(
                                            'observacoes', StringType(), True
                                        ),
                                        StructField(
                                            'secao',
                                            StructType(
                                                [
                                                    StructField(
                                                        'id',
                                                        StringType(),
                                                        True,
                                                    ),
                                                    StructField(
                                                        'descricao',
                                                        StringType(),
                                                        True,
                                                    ),
                                                    StructField(
                                                        'observacoes',
                                                        StringType(),
                                                        True,
                                                    ),
                                                ]
                                            ),
                                            True,
                                        ),
                                    ]
                                ),
                                True,
                            ),
                        ]
                    ),
                    True,
                ),
            ]
        )
