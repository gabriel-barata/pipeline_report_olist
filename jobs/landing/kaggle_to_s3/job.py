from pathlib import Path

from jobs.base.job import Job

BASE_LOCAL_PATH = (
    Path(__file__).parent.parent.parent.parent / 'resources' / 'sample_data'
)
BASE_S3_PATH = 's3a://olist-datalake-prd-us-east-1-269012942764-landing/kaggle/'

FILES = {
    'order_items': 'olist_order_items_dataset.csv',
    'order_payments': 'olist_order_payments_dataset.csv',
    'products': 'olist_products_dataset.csv',
}


class KaggleToS3Job(Job):
    def run(self) -> None:
        for table_name, file_name in FILES.items():
            file_path = str(BASE_LOCAL_PATH / file_name)
            output_path = BASE_S3_PATH + table_name

            data = self.spark.read.csv(
                file_path, sep=',', header=True, inferSchema=True
            )

            data = self.generate_partition_columns(data)

            (
                data.write.mode('overwrite')
                .partitionBy('YEAR', 'MONTH', 'DAY')
                .parquet(output_path)
            )
