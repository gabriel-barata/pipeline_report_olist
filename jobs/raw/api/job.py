from pyspark.sql import DataFrame
from pyspark.sql.functions import col
from pyspark.sql.types import StringType, StructField, StructType

from jobs.base.job import Job
from jobs.base.variables import LANDING_BUCKET, RAW_BUCKET


class ApiJob(Job):
    source = LANDING_BUCKET + 'api_CNAE/'
    target = RAW_BUCKET + 'api_CNAE/'

    def run(self) -> None:
        data = self.spark.read.json(self.source, schema=self.schema)
        data = self.filter_by_ref_date(data)
        data = self.flatten_dataframe(data)

        (
            data.write.mode('overwrite')
            .partitionBy('YEAR', 'MONTH', 'DAY')
            .parquet(self.target)
        )

    @property
    def schema(self) -> StructType:
        return StructType(
            [
                StructField('id', StringType(), True),
                StructField('descricao', StringType(), True),
                StructField('observacoes', StringType(), True),
                StructField(
                    'grupo',
                    StructType(
                        [
                            StructField('id', StringType(), True),
                            StructField('descricao', StringType(), True),
                            StructField('observacoes', StringType(), True),
                            StructField(
                                'divisao',
                                StructType(
                                    [
                                        StructField('id', StringType(), True),
                                        StructField(
                                            'descricao', StringType(), True
                                        ),
                                        StructField(
                                            'observacoes', StringType(), True
                                        ),
                                        StructField(
                                            'secao',
                                            StructType(
                                                [
                                                    StructField(
                                                        'id',
                                                        StringType(),
                                                        True,
                                                    ),
                                                    StructField(
                                                        'descricao',
                                                        StringType(),
                                                        True,
                                                    ),
                                                    StructField(
                                                        'observacoes',
                                                        StringType(),
                                                        True,
                                                    ),
                                                ]
                                            ),
                                            True,
                                        ),
                                    ]
                                ),
                                True,
                            ),
                        ]
                    ),
                    True,
                ),
            ]
        )

    @staticmethod
    def flatten_dataframe(df: DataFrame) -> DataFrame:
        return df.select(
            col('id'),
            col('descricao'),
            col('observacoes'),
            col('grupo.id').alias('grupo_id'),
            col('grupo.descricao').alias('grupo_descricao'),
            col('grupo.observacoes').alias('grupo_observacoes'),
            col('grupo.divisao.id').alias('divisao_id'),
            col('grupo.divisao.descricao').alias('divisao_descricao'),
            col('grupo.divisao.observacoes').alias('divisao_observacoes'),
            col('grupo.divisao.secao.id').alias('secao_id'),
            col('grupo.divisao.secao.descricao').alias('secao_descricao'),
            col('grupo.divisao.secao.observacoes').alias('secao_observacoes'),
            col('YEAR'),
            col('MONTH'),
            col('DAY'),
        )
