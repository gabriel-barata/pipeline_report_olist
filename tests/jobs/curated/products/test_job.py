from unittest.mock import PropertyMock, patch

import pytest
from pyspark.sql import DataFrame, SparkSession

from jobs.curated.products.job import ProductsJob


@pytest.fixture(scope='module')
def products_job(spark: SparkSession) -> ProductsJob:
    return ProductsJob(
        spark=spark, is_test=True, config={'ref_date': '2025-05-22'}
    )


def test_job(
    products_job: ProductsJob,
    source_order_items: DataFrame,
    source_products: DataFrame,
    snapshot,
) -> None:

    with (
        patch.object(
            ProductsJob, 'source_products', new_callable=PropertyMock
        ) as mock_products,
        patch.object(
            ProductsJob, 'source_order_items', new_callable=PropertyMock
        ) as mock_order_items,
    ):

        mock_order_items.return_value = source_order_items
        mock_products.return_value = source_products

        output = products_job.run()

        assert output.collect() == snapshot
