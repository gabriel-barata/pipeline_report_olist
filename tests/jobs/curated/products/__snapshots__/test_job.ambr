# serializer version: 1
# name: test_job
  list([
    Row(
      '4244733e06e7ecb4970a6e2683c13e61',
      'perfumaria',
      40,
      287,
      1,
      225,
      16,
      10,
      'baixo',
      '2025',
      '05',
      '22',
    ),
  ])
# ---
